File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int main(void)
{
    while (true)
    {
        long number = get_long("Number: ");
        long _number = number;
        int count = 0;
        int *array = malloc(sizeof(int) * 16);
        while (true)
        {
            array[count] = _number % 10;
            _number = _number / 10;
            count++;
            if (_number <= 0)
            {
                break;
            }
        }

        if (count < 13 || count > 16)
        {
            printf("INVALID: out of range.\n");
            return -1;
        }

        int sum = 0;
        for (int i = 0; i < count; i++)
        {
            if (i % 2 == 0)
            {
                sum += array[i];
            }
            else
            {
                int _mult = array[i] * 2;
                if (_mult > 9)
                {
					sum += _mult - 9;
                }
                else
                {
                    sum += _mult;
                }
            }
        }

        if (sum % 10 == 0)
        {
            if (array[count - 1] == 3)
            {
                if (array[count - 2] == 4 || array[count - 2] == 7)
                {
                    printf("AMEX\n");
                    return 0;
                }
            }
            else if (array[count - 1] == 5)
            {
                if (array[count - 2] <= 5)
                {
                    printf("MASTERCARD\n");
                    return 0;
                }
            }
            else if (array[count - 1] == 4)
            {
                printf("VISA\n");
                return 0;
            }
        }
        printf("INVALID: bad check sum.\n");
        return -2;
    }
}

File: pset6/credit/credit.py
from cs50 import get_int

def list_number(number):
    array = []
    while True:
        array.append(number % 10)
        number = number // 10
        if number <= 0:
            break
    return array[::-1]


def is_amex(list_number):
    if list_number[0] == 3:
        if list_number[1] == 4 or list_number[1] == 7:
            return True
    return False


def is_master(list_number):
    if list_number[0] == 5:
        if list_number[0] <= 5:
            return True
    return False


def is_visa(list_number):
    if list_number[0] == 4:
        return True
    return False


while True:
    data = get_int("Number: ")
    numbers = list_number(data)
    reverse_numbers = numbers[::-1]
    numbers_length = len(numbers)

    if numbers_length < 13 or numbers_length > 16:
        print("INVALID")
        break

    sum = 0
    for i in reverse_numbers[1::2]:
        sum += i * 2
        if i >= 5:
            sum -= 9

    for i in reverse_numbers[0::2]:
        sum += i

    if sum % 10 == 0:
        if is_amex(numbers):
            print("AMEX")
        elif is_master(numbers):
            print("MASTERCARD")
        elif is_visa(numbers):
            print("VISA")
        else:
            print("INVALID")
        break
